web-app/src/common/graphWinPrediction.ts

```diff
@@ -0,0 +1,131 @@
+    row.forEach((_cell, x) => {
+      if (_cell.value === stoneColor || _cell.value === "empty") {
+        addVertex(hexBoardWinPredictionGraph, `${y}-${x}`);
+      }
+    });
+  });
+}
+
+function addVertex(hexBoardWinPredictionGraph: HexBoardWinPredictionGraph, id: string) {
+  hexBoardWinPredictionGraph.vertexMap.set(id, hexBoardWinPredictionGraph.winPrediction.addVertex(id));
+}
+
+function createPlayableCellEdges(
+  gameState: GameState,
+  hexBoardWinPredictionGraph: HexBoardWinPredictionGraph,
+  stoneColor: StoneColor,
+) {
+  gameState.board.forEach((row, y) => {
+    row.forEach((cell, x) => {
+      if (cell.value == stoneColor || cell.value == "empty") {
+        const currentCell: Coordinates = { y: y, x: x };
+        const directNeighborCells = [
+          { y: y, x: x - 1 }, { y: y, x: x + 1 },
+          { y: y - 1, x: x }, { y: y + 1, x: x },
+          { y: y - 1, x: x - 1 }, { y: y + 1, x: x + 1 },
+        ];
+        directNeighborCells.forEach(neighbor => {
+          if (!doesCellExist(gameState, neighbor) || doesCellExistAndHaveStone(gameState, neighbor, stoneColor === "black" ? "white" : "black")) return;
+          const cost = doesCellExistAndHaveStone(gameState, neighbor, stoneColor) ? LOW_EDGE_COST : HIGH_EDGE_COST;
+          addEdgeWithCost(hexBoardWinPredictionGraph, `${currentCell.y}-${currentCell.x}`, `${neighbor.y}-${neighbor.x}`, cost);
+        })
+      }
+    })
+  });
+}
+
+/**
+ * Create edges for start and end nodes of the given color.
+ * White color borders are top-bottom and black color border are left-right.
+ */
+function createStartEndEdges(
+  gameState: GameState,
+  hexBoardWinPredictionGraph: HexBoardWinPredictionGraph,
+  stoneColor: StoneColor,
+) {
+  gameState.board.forEach((_, i) => {
+    if (stoneColor === "black") {
+      const currentFirstColumnCellValue = gameState.board[i][0].value;
+      if (currentFirstColumnCellValue !== "white") {
+        addEdgeWithCost(hexBoardWinPredictionGraph, BLACK_NODE_START, `${i}-0`, currentFirstColumnCellValue === "black" ? LOW_EDGE_COST : HIGH_EDGE_COST);
+      }
+      const currentLastColumnCellValue = gameState.board[i][gameState.board.length - 1].value;
+      if (currentLastColumnCellValue !== "white") {
+        addEdgeWithCost(hexBoardWinPredictionGraph, `${i}-${gameState.board.length - 1}`, BLACK_NODE_END, UNDEFINED_EDGE_COST);
+      }
+    }
+    else {
+      const currentFirstRowCellValue = gameState.board[0][i].value;
+      if (currentFirstRowCellValue !== "black") {
+        addEdgeWithCost(hexBoardWinPredictionGraph, WHITE_NODE_START, `0-${i}`, currentFirstRowCellValue === "white" ? LOW_EDGE_COST : HIGH_EDGE_COST);
+      }
+      const currentLastRowCellValue = gameState.board[gameState.board.length - 1][i].value;
+      if (currentLastRowCellValue !== "black") {
+        addEdgeWithCost(hexBoardWinPredictionGraph, `${gameState.board.length - 1}-${i}`, WHITE_NODE_END, UNDEFINED_EDGE_COST);
+      }
+    }
+  });
```

Comment:
I get what you did there, but it is a bit hard to read and understand imo.

One way you could have improved this (but I'm not sure it's worth the refactoring though), is to consider this:
- For now, your function `addEdgeWithCost()` simply does what it's told, ie. it creates an edge with a cost.
- But I believe it could do more than that, since the behaviour is always the same:
  - Create the edge only if the starting node exists and is not white
  - Create the edge only if the end node exists and is not white
  - Compute the cost depending on the end node: if empty, HIGH COST, otherwise, LOW COST
- with that in mind, you could delegate all these checks to the child function, and then your `createStartEndEdges()` function would be much simpler

Reply:
Yes, I agree with you, that part is not very readable and probably needs to be reworked. But I'm not sure I understand your solution because addEdgeWithCost is also used to draw edges between common cells so I think it's good to let it do the minimal work without adding more complexity to it.

Reply:
Unless I'm missing something, even in the case of 'common cells', the function as I describe it would still work, wouldn't it?

Anyways, as I said,  I'm not sure it's worth the refactoring for so little. Let's keep it that way for now.